redis/src/main/java/org/hibernate/ogm/datastore/redis/AbstractRedisDialect.java

```diff
@@ -316,4 +366,119 @@ protected void storeAssociation(
 			}
 		}
 	}
+
+	/**
+	 * Scan over keys. This method is aware whether the the client is connected to a Redis Cluster.
+	 * If so, then a Redis Cluster scan requires to iterate over master nodes and keep the
+	 * state within a custom {@link KeyScanCursor} instance.
+	 *
+	 * @param cursor
+	 * @param scanArgs
+	 *
+	 * @return
+	 */
+	protected KeyScanCursor<String> scan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		if ( !clusterMode ) {
+			return scan( connection, cursor, scanArgs );
+		}
+
+		return clusterScan( cursor, scanArgs );
+	}
+
+	private KeyScanCursor<String> scan(
+			RedisClusterCommands<String, String> commands,
+			KeyScanCursor<String> cursor,
+			ScanArgs scanArgs) {
+		if ( cursor != null ) {
+			return commands.scan( cursor, scanArgs );
+		}
+		return commands.scan( scanArgs );
+	}
+
+	@SuppressWarnings("unchecked")
+	private KeyScanCursor<String> clusterScan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
```

Comment:
So scanning across all the masters sounds awefully inperformant :) Maybe we can find a way to limit the number of master nodes one entity type is sharded across, so we can limit the scan to those? Another consideration would be parallelization of this scan (e.g. different master nodes could be queried in parallel).